osbuild.py

```diff
@@ -86,91 +88,30 @@ def unmount(self):
-
-        binds = [f"{tree}:/run/osbuild/tree", "/dev:/dev"]
-
-        try:
-            r = self.run([f"/run/osbuild/{name}"],
-                binds=binds,
-                readonly_binds=robinds,
-                input=json.dumps(args),
-                encoding="utf-8",
-                stdout=subprocess.PIPE if not interactive else None,
-                stderr=subprocess.STDOUT if not interactive else None,
-                check=True)
-        except subprocess.CalledProcessError as error:
-            raise StageFailed(name, error.returncode, error.stdout)
-
-        return {
-            "name": name,
-            "output": r.stdout
-        }
-
-    def run_assembler(self, assembler, tree, output_dir=None, interactive=False):
-        if output_dir and not os.path.exists(output_dir):
-            os.makedirs(output_dir)
-
-        name = assembler["name"]
-        args = {
-            "tree": "/run/osbuild/tree",
-            "options": assembler.get("options", {}),
-        }
-        robinds = [
-            f"{tree}:/run/osbuild/tree",
-            f"{libdir}/assemblers/{name}:/run/osbuild/{name}"
-        ]
-        robinds.extend(self._get_system_resources_from_etc(assembler))
-        binds = ["/dev:/dev"]
-
-        if output_dir:
-            binds.append(f"{output_dir}:/run/osbuild/output")
-            args["output_dir"] = "/run/osbuild/output"
-
+    def run(self, name, args, binds=[], readonly_binds=[], interactive=False):
         try:
-            r = self.run([f"/run/osbuild/{name}"],
-                binds=binds,
-                readonly_binds=robinds,
-                input=json.dumps(args),
-                encoding="utf-8",
-                stdout=subprocess.PIPE if not interactive else None,
-                stderr=subprocess.STDOUT if not interactive else None,
-                check=True)
+            r = subprocess.run([
+                "systemd-nspawn",
+                "--quiet",
+                "--as-pid2",
+                "--link-journal=no",
+                "--volatile=yes",
+                "--property=DeviceAllow=/dev/loop-control rw",
+                "--property=DeviceAllow=block-loop rw",
+                "--property=DeviceAllow=block-blkext rw",
+                f"--machine={self.machine_name}",
+                f"--directory={self.root}",
+                f"--bind={libdir}/osbuild-run:/run/osbuild/osbuild-run",
+                *[f"--bind={b}" for b in binds],
+                *[f"--bind-ro={b}" for b in readonly_binds],
+                "/run/osbuild/osbuild-run",
+                f"/run/osbuild/{name}",
+            ],
+            input=json.dumps(args),
+            encoding="utf-8",
+            stdout=subprocess.PIPE if not interactive else None,
+            stderr=subprocess.STDOUT if not interactive else None,
+            check=True)
         except subprocess.CalledProcessError as error:
             raise StageFailed(name, error.returncode, error.stdout)
```

Comment:
I like the distinction between `BuildRoot.run()`, which ran arbitrary commands in the build root, and (now) `Stage.run()`, which runs a stage in the build root. To be honest, I'm not sure if we need the former. I just liked it.

Do we want to move the exception conversion into the other functions and split them into `StageFailed` and `AssemblerFailed`?

Reply:
Totally agreed.